File: src/cdoan/stanfordalgo/week1/MergeSort.java
package cdoan.stanfordalgo.week1;

import java.util.Random;

public class MergeSort {

  /**
   * Test sort method using a random integer array as input.
   */
  public static void main(String[] args) {
    for (int test = 0; test < 1300; ++test) {
      Integer[] a = randomIntegerArr();
      sort(a, new Integer[a.length]);
      //assert isSorted(a);
      if (!isSorted(a)) {
        throw new AssertionError("The array is Not sorted!");
      }
    }
  }

  /* Utility method for preparing test data */
  private static Integer[] randomIntegerArr() {
    Random random = new Random();
    int len = 6 + random.nextInt(11) + random.nextInt(21);
    Integer[] result = new Integer[len];
    for (int i = 0; i < len; i++) {
      result[i] = random.nextInt(200) - 100;
    }
    return result;
  }

  /**
   * Sort array a using array aux as buffer.
   */
  public static <T extends Comparable<T>> void sort(T[] a, T[] aux) {
    if (a == null || aux == null) {
      throw new NullPointerException("The input array or auxiliary array is null!");
    }
    if (aux == a) {
      throw new IllegalArgumentException("Auxiliary array mustn't be input array!");
    }
    if (aux.length < a.length) {
      throw new IllegalArgumentException("Auxiliary array mustn't be shorter than input array!");
    }

    final int n = a.length;
    System.arraycopy(a, 0, aux, 0, n);
    sort(a, 0, n - 1, aux);
  }

  /* Recursively sort a[lo..hi] using aux as buffer */
  private static <T extends Comparable<T>> void sort(T[] a, int lo, int hi, T[] aux) {
    // base case
    if (lo >= hi) {
      return;
    }

    int mid = (lo + hi) / 2;
    sort(aux, lo, mid, a);
    sort(aux,mid + 1, hi, a);
    merge(aux, lo, mid, hi, a);
  }

  /* Merge sorted aux[lo..mid] and aux[mid+1..hi] into a[lo..hi] */
  private static <T extends Comparable<T>> void merge(T[] aux, int lo, int mid, int hi, T[] a) {
    int i = lo;
    int j = mid + 1;
    int k = lo;
    while (i <= mid && j <= hi) {
      if (less(aux[i], aux[j])) {
        a[k++] = aux[i++];
      } else {
        a[k++] = aux[j++];
      }
    }
    while (i <= mid) {
      a[k++] = aux[i++];
    }
    while (j <= hi) {
      a[k++] = aux[j++];
    }
  }
  
  
  /* Helper method for testing if v is less than w */
  private static <T extends Comparable<T>> boolean less(T v, T w) {
    return v.compareTo(w) < 0;
  }
  
  /* Helper method for testing if given array is sorted */
  private static <T extends Comparable<T>> boolean isSorted(T[] a) {
    final int n = a.length;
    for (int i = 1; i < a.length; ++i) {
      if (less(a[i], a[i - 1])) {
        return false;
      }
    }
    return true;
  }
  
}

File: src/cdoan/stanfordalgo/week1/MergeSort.kt
package cdoan.stanfordalgo.week1

import cdoan.util.ArrayUtil
import java.util.Random

/**
 * Created by co on 14/06/2017.
 */

object MergeSort {

    @JvmStatic fun main(args: Array<String>) {
        //val a = intArrayOf(8, 6, 1, 7, 5)
        //val a = intArrayOf(6, 2, 8, 1, 7, 3, 5)
        val a = ArrayUtil.randomArr()
        val acopy = IntArray(a.size)
        System.arraycopy(a, 0, acopy, 0, a.size)
        ArrayUtil.printArr(a)
        sort(a)
        ArrayUtil.printArr(a)
        println("array is sorted: ${ArrayUtil.isArraySorted(a)}")

        println()

        ArrayUtil.printArr(acopy)
        sortBottomUp(acopy)
        ArrayUtil.printArr(acopy)
        println("array is sorted: ${ArrayUtil.isArraySorted(acopy)}")
    }

    fun sortBottomUp(a: IntArray?) {
        if (a == null || a.size <= 1) {
            return
        }

        val n = a.size
        val buff = IntArray(n)
        System.arraycopy(a, 0, buff, 0, n)
        val numPasses = numDivideBy2ToReach1(n)
        var pass = 0
        var size = 1
        while (pass < numPasses) {
            var lo = 0
            var mid = lo + size - 1
            var hi = mid + size
            while (hi < n) {
                if (pass % 2 == 0) {
                    if (numPasses % 2 == 0) {
                        // NOTE: start merging from a into buff
                        merge(a, lo, mid, hi, buff) //merge from a into buff
                    } else {
                        // NOTE: start merging from buff into a
                        merge(buff, lo, mid, hi, a) //merge from buff into a
                    }
                } else {
                    if (numPasses % 2 == 0) {
                        // NOTE: start merging from buff into a
                        merge(buff, lo, mid, hi, a) //merge from buff into a
                    } else {
                        // NOTE: start merging from a into buff
                        merge(a, lo, mid, hi, buff) //merge from a into buff
                    }
                }
                lo = hi + 1
                mid = lo + size - 1
                hi = mid + size
            }

            if (lo < n) {
                // handle the last n-lo elements
                if (n - lo > size) {
                    mid = lo + size - 1
                    hi = n - 1
                    if (pass % 2 == 0) {
                        if (numPasses % 2 == 0) {
                            merge(a, lo, mid, hi, buff) //merge from a into buff
                        } else {
                            merge(buff, lo, mid, hi, a) //merge from buff into a
                        }
                    } else {
                        if (numPasses % 2 == 0) {
                            merge(buff, lo, mid, hi, a) //merge from buff into a
                        } else {
                            merge(a, lo, mid, hi, buff) //merge from a into buff
                        }
                    }
                }
            }
            pass++
            size *= 2
        }
    }

    private fun numDivideBy2ToReach1(n: Int): Int {
        var n = n
        var logOfN = 0
        val origin = n
        while (n > 1) {
            n /= 2
            logOfN++
        }
        return if (1 shl logOfN == origin) logOfN else logOfN + 1
    }

    fun sort(a: IntArray?) {
        if (a == null || a.size <= 1) {
            return
        }

        val buff = IntArray(a.size)
        System.arraycopy(a, 0, buff, 0, a.size)
        sort(a, 0, a.size - 1, buff)
    }

    /**
     * Sort a[lo to hi] in-place using buffer 'buff'
     */
    private fun sort(a: IntArray, lo: Int, hi: Int, buff: IntArray) {
        // base case
        if (lo >= hi)
            return

        val mid = (lo + hi) / 2
        sort(buff, lo, mid, a)
        sort(buff, mid + 1, hi, a)
        merge(buff, lo, mid, hi, a)
    }

    /**
     * Merge sorted buff[lo to mid] and buff[mid+1 to hi] into a[lo to hi]
     */
    private fun merge(buff: IntArray, lo: Int, mid: Int, hi: Int, a: IntArray) {
        var i = lo
        var j = mid + 1
        var k = lo
        while (i <= mid && j <= hi) {
            if (buff[i] < buff[j]) {
                a[k++] = buff[i++]
            } else {
                a[k++] = buff[j++]
            }
        }
        while (i <= mid) {
            a[k++] = buff[i++]
        }
        while (j <= hi) {
            a[k++] = buff[j++]
        }
    }

}
